batch_job/src/batch_job_amazon.c

```diff
@@ -176,6 +171,68 @@ static struct jx * aws_describe_instance( struct aws_config *c, const char *inst
 	return j;
 }
 
+/*
```

Comment:
If I am reading this correctly, what you are doing is asking AWS to put labels on each of the machines to describe their state.  I don't think this is the right way to go about it, since those queries and modifications can be quite expensive.

I think a much easier way to go about it would be to simply keep a local list of "available" VMs.  Whenever a job completes, put the VM on the list.  Whenever a new job is submitted, look for an available VM on the list.  If one is available/compatible, then use it.  Otherwise, create a new one.